pkg/testing/tools/artifacts_snapshot.go

```diff
@@ -0,0 +1,204 @@
+	// heartbeat-8.10.0-SNAPSHOT-linux-x86_64.tar.gz
+	Name       string
+	TargetPath string
+}
+
+func NewArtifactSnapshotClient() *ArtifactSnapshotClient {
+	return &ArtifactSnapshotClient{
+		url:        defaultAPIURL,
+		httpClient: new(http.Client),
+	}
+}
+
+func (c *ArtifactSnapshotClient) DownloadPackages(ctx context.Context, packageRequests []PackageRequest, version string) error {
+	manResponse, err := c.getManifestResponse(ctx, version)
+	if err != nil {
+		return fmt.Errorf("getting manifest response: %w", err)
+	}
+	manifestPackages, err := c.getManifestPackages(ctx, manResponse.ManifestURL)
+	if err != nil {
+		return fmt.Errorf("getting manifest packages: %w", err)
+	}
+	for _, pkg := range packageRequests {
+		pkgData, ok := manifestPackages[pkg.Name]
+		if !ok {
+			return fmt.Errorf("package %s not found in manifest", pkg)
+		}
+		err := c.downloadPackage(ctx, pkgData, pkg.TargetPath)
+		if err != nil {
+			return fmt.Errorf("downloading package: %w", err)
+		}
+	}
+	return nil
+}
+
+func (c *ArtifactSnapshotClient) getManifestResponse(ctx context.Context, version string) (*ArtifactsManifestResponse, error) {
+	endpoint := fmt.Sprintf(beatsManifestEndpoint, version)
+	url := fmt.Sprintf("%s/%s", c.url, endpoint)
+	response, err := c.createAndPerformRequest(ctx, url)
+	if err != nil {
+		return nil, fmt.Errorf("creating and performing request: %w", err)
+	}
+	manResponse, err := checkResponseAndUnmarshal[ArtifactsManifestResponse](response)
+	if err != nil {
+		return nil, fmt.Errorf("checking response and unmarshalling: %w", err)
+	}
+	return manResponse, nil
+}
+
+func (c *ArtifactSnapshotClient) getManifestPackages(ctx context.Context, manifestURL string) (map[string]PackageData, error) {
+	response, err := c.createAndPerformRequest(ctx, manifestURL)
+	if err != nil {
+		return nil, fmt.Errorf("creating and performing request: %w", err)
+	}
+	manResponse, err := checkResponseAndUnmarshal[Projects](response)
+	if err != nil {
+		return nil, fmt.Errorf("checking response and unmarshalling: %w", err)
+	}
+	allPackages := make(map[string]PackageData)
+	for _, v := range maps.Values(manResponse.Projects) {
+		for k, pack := range v.Packages {
+			allPackages[k] = pack
+		}
+	}
+	return allPackages, nil
+}
+
+func (c *ArtifactSnapshotClient) downloadPackage(ctx context.Context, pkg PackageData, destinationDir string) error {
+
+	return nil
+}
+
+func downloadFile(url string, path string) error {
+	var filePath string
+	if path == "" {
+		tempParentDir := filepath.Join(os.TempDir(), uuid.NewString())
+		filePath = filepath.Join(tempParentDir, uuid.NewString())
+		path = filePath
+	} else {
+		filePath = filepath.Join(path, uuid.NewString())
+	}
+
+	tempFile, err := os.Create(filePath)
+	if err != nil {
+		return fmt.Errorf("creating file: %w", err)
+	}
+	defer tempFile.Close()
+	exp := getExponentialBackOff(3)
+
+	retryCount := 1
+	var fileReader io.ReadCloser
+	download := func() error {
+		resp, err := http.Get(url)
```

Comment:
Why isn't it a method of `ArtifactSnapshotClient`? If it were, you could use the http client `ArtifactSnapshotClient` has as you did on `(aac *ArtifactSnapshotClient) createAndPerformRequest` or even use `createAndPerformRequest` directly.
It'd be better to use that HTTP client as well as having 1 single way of making HTTP requests, all using the same HTTP client